### ISSUE TITLE
`getattr(self.settings, "build_type", "")` should return "" instead of raising ConanException

### ISSUE THREAD

**Issue by DmitrySokolov** [conan-io/conan#4562]
When `build_type` is `None` the code `getattr(self.settings, "build_type", "")` raises exception `ConanException`. Which break the default behavior:

https://docs.python.org/3/library/functions.html#getattr
> **getattr(object, name[, default])**
>
>    Return the value of the named attribute of object. name must be a string. If the string is the name of one of the object’s attributes, the result is the value of that attribute. For example, `getattr(x, 'foobar')` is equivalent to `x.foobar`. If the named attribute does not exist, default is returned if provided, otherwise AttributeError is raised.

**Comment by memsharded** [comment]
Hi @DmitrySokolov 

I am having a look, and I have identified an error when that returned ``SettingsItem`` object is tried to be converted to ``str``, that raises. I would like to ask:
- What is the context you are doing that? Why aren't you doing just ``self.settings.build_type``?
- What is the ConanException and message that it produces?
- Are you defining ``build_type`` as setting in the recipe, then not giving it a value (None), or is it that you are not defining ``build_type`` at all in the recipe?

**Comment by DmitrySokolov** [comment]
v1

conanfile.py
```py
import os
from conans import ConanFile, tools


class Test1Conan(ConanFile):
    name = "Test1"
    version = "1.0"
    settings = "os", "compiler", "build_type", "arch"
    description = "<Description of Test1 here>"
    url = "None"
    license = "None"
    author = "None"
    topics = None

    def configure(self):
        if str(self.settings.os).lower() not in ["windows", "macos"]:
            del self.settings.build_type

    def package(self):
        self.output.info("Creating Test1.conf")
        conf_file = os.path.join(self.package_folder, "Test1.conf")
        tools.save(conf_file, f"[Paths]\n"
                              f"Prefix = ..\n"
                              f"OS = {str(self.settings.os)}\n"
                              f"Arch = {str(self.settings.arch)}\n"
                              f"CC = {str(self.settings.compiler)}\n"
                              f"Type = {str(self.settings.build_type)}\n")
        self.output.info("  " + conf_file)
```

exporting
```dos
C:\Dev\Tools\Python3\python.exe C:/Dev/Tools/Python3/Lib/site-packages/conans/conan.py export-pkg . tester/testing -s os=Android -s os.api_level=16 -s arch=x86 -s compiler=clang -s compiler.version=8.0 -s compiler.libcxx=libstdc++ -s build_type=Release -f
Exporting package recipe
Test1/1.0@tester/testing: A new conanfile.py version was exported
Test1/1.0@tester/testing: Folder: D:\Users\Dmitry\.conan\data\Test1\1.0\tester\testing\export
Packaging to 87d2b4aa6bf7608e9193c4dc2557fd1521ded62b
Test1/1.0@tester/testing: Generating the package
Test1/1.0@tester/testing: Package folder D:\Users\Dmitry\.conan\data\Test1\1.0\tester\testing\package\87d2b4aa6bf7608e9193c4dc2557fd1521ded62b
Test1/1.0@tester/testing: Calling package()
Test1/1.0@tester/testing: Creating Test1.conf
ERROR: Test1/1.0@tester/testing: Error in package() method, line 23
	tools.save(conf_file, f"[Paths]\n"
	ConanException: 'settings.build_type' doesn't exist
'settings' possible configurations are ['arch', 'compiler', 'os']
```

**Comment by DmitrySokolov** [comment]
v2

conanfile.py
```py
import os
from conans import ConanFile, tools


class Test1Conan(ConanFile):
    name = "Test1"
    version = "1.0"
    settings = "os", "compiler", "build_type", "arch"
    description = "<Description of Test1 here>"
    url = "None"
    license = "None"
    author = "None"
    topics = None

    def configure(self):
        if str(self.settings.os).lower() not in ["windows", "macos"]:
            del self.settings.build_type

    def package(self):
        self.output.info("Creating Test1.conf")
        conf_file = os.path.join(self.package_folder, "Test1.conf")
        bt = getattr(self.settings, "build_type", None)
        tools.save(conf_file, f"[Paths]\n"
                              f"Prefix = ..\n"
                              f"OS = {str(self.settings.os)}\n"
                              f"Arch = {str(self.settings.arch)}\n"
                              f"CC = {str(self.settings.compiler)}\n"
                              f"Type = {str(bt)}\n")
        self.output.info("  " + conf_file)
```

exporting
```dos
C:\Dev\Tools\Python3\python.exe C:/Dev/Tools/Python3/Lib/site-packages/conans/conan.py export-pkg . tester/testing -s os=Android -s os.api_level=16 -s arch=x86 -s compiler=clang -s compiler.version=8.0 -s compiler.libcxx=libstdc++ -s build_type=Release -f
Exporting package recipe
Test1/1.0@tester/testing: A new conanfile.py version was exported
Test1/1.0@tester/testing: Folder: D:\Users\Dmitry\.conan\data\Test1\1.0\tester\testing\export
Packaging to 87d2b4aa6bf7608e9193c4dc2557fd1521ded62b
Test1/1.0@tester/testing: Generating the package
Test1/1.0@tester/testing: Package folder D:\Users\Dmitry\.conan\data\Test1\1.0\tester\testing\package\87d2b4aa6bf7608e9193c4dc2557fd1521ded62b
Test1/1.0@tester/testing: Calling package()
Test1/1.0@tester/testing: Creating Test1.conf
ERROR: Test1/1.0@tester/testing: Error in package() method, line 22
	bt = getattr(self.settings, "build_type", None)
	ConanException: 'settings.build_type' doesn't exist
'settings' possible configurations are ['arch', 'compiler', 'os']
```

**Comment by DmitrySokolov** [comment]
v3

conanfile.py
```py
import os
from conans import ConanFile, tools


class Test1Conan(ConanFile):
    name = "Test1"
    version = "1.0"
    settings = "os", "compiler", "build_type", "arch"
    description = "<Description of Test1 here>"
    url = "None"
    license = "None"
    author = "None"
    topics = None

    def package_id(self):
        if str(self.info.settings.os).lower() in ["windows", "macos"]:
            self.info.settings.build_type = "ANY"
            if self.info.settings.compiler == "Visual Studio":
                self.info.settings.compiler.runtime = "ANY"

    def package(self):
        self.output.info("Creating Test1.conf")
        conf_file = os.path.join(self.package_folder, "Test1.conf")
        tools.save(conf_file, f"[Paths]\n"
                              f"Prefix = ..\n"
                              f"OS = {str(self.info.settings.os)}\n"
                              f"Arch = {str(self.info.settings.arch)}\n"
                              f"CC = {str(self.info.settings.compiler)}\n"
                              f"Type = {str(self.info.settings.build_type)}\n")
        self.output.info("  " + conf_file)
```

**Comment by DmitrySokolov** [comment]
I think you'll propose to use v3?

**Comment by DmitrySokolov** [comment]
I use a sort of `PackageBase` class (via `python_requires`). Is `self.info` available in `config_options()` and in `configure()`?

**Comment by DmitrySokolov** [comment]
However, I see no reason why `getattr(o, name, default)` should break the contract and raise an exception instead of returning a value.

**Comment by memsharded** [comment]
Hi @DmitrySokolov 

No, I am not going to propose v3, please don't use ``self.info`` for this, that is intended just for ``package_id()`` and its value can be different to the actual settings that have been defined.

> However, I see no reason why `getattr(o, name, default)` should break the contract and raise an exception instead of returning a value.

This is by design. The reason `getattr(o, name, default)` changes the default python behavior is because the ``getattr()`` method is actually the way it is possible to implement the ``self.settings.build_type``,and in general a ``self.settings.xxxxx`` without ``xxxxxx`` being a real attribute. It isn't, it is generated on the fly based on the settings.yml and other info. It is used to force difference things, like for example exact matching or failing. You know, comparing ``self.settings.build_type == "random-string"`` will also raise an error. This is for sure not the default behavior of a comparison operator.

These deviations from the default behavior makes sense because we are not writing python programs with it. We are writing conan package recipes, a DSL based on python. In this case it makes more sense to have a robust and helpful UI that fails very early on failures (like typos ``self.settings.buildtype``, or ``Releseae``), than to adhere to that python default behavior. Unfortunately there is no way it could be respected when called explicitly with a default attribute, the override is only called with the attribute name, but not the default, so it cannot do nothing but fail (i.e. the implementation cannot know if it is being called with ``self.settings.build_type`` or ``getattr(self.settings, "build_type", None)``

To summarize: it would be good to keep that default if possible, but it isn't possible to implement it and keep the ``self.settings.xxxx`` current behavior, so this can't be done.


I think you might be interested in the ``self.settings.get_safe("build_type")`` function, that will return None if it is not defined. Good thing is that it also works with compound settings: ``self.settings.get_safe("compiler.version")`` is safe to use, and will return None both if compiler or version aren't defined. It is not documented, but if you feel it safe, we could document it and make it part of the public and stable api.

**Comment by DmitrySokolov** [comment]
Oh, I see.

I didn't catch why not to use `self.info`: if I'll use v3 then if I'll
```dos
conan install Test1/1.0@tester/testing -s os=Windows -s arch=x86 -s compiler="Visual Studio" -s compiler.version=15 -s compiler.toolset=v141 -s compiler.runtime=MD -s build_type=Release

conan install Test1/1.0@tester/testing -s os=Windows -s arch=x86 -s compiler="Visual Studio" -s compiler.version=15 -s compiler.toolset=v141 -s compiler.runtime=MDd -s build_type=Debug
```
I'll get the same package for both `Release` and `Debug` versions (what was the purpose).
So, why not?

**Comment by DmitrySokolov** [comment]
I see the drawback in that I have to always specify `-s build_type=X` even if `build_type` is `ANY` for all platforms. And in case of
```py
    def package_id(self):
        self.info.header_only()
```
it will be very annoying.

**Comment by memsharded** [comment]
Yes, exactly, if it is a header-only, it makes sense to remove the ``build_type`` not to define that it doesn't affect the package-ID, those are different concepts. The first one means that there is no change in the package if you change the ``build_type``, the second one means that a package that is built with either Debug or Release will be compatible with the other build_type. But the system will only manage one binary for both configurations.

So please, tell me if the ``self.settings.get_safe("build_type")`` makes sense for you, and we will considering adding it to the public api, or if the rationale of the behaviors is clear and the issue could be closed. Thanks!

**Comment by DmitrySokolov** [comment]
`get_safe()` suits me well. Thank you!

**Comment by memsharded** [comment]
``get_safe()`` has been documented and added unittest in https://github.com/conan-io/conan/pull/4571, so it will be part of the public api in conan 1.13, so I'd say we could close this issue now, but please ask any further question if necessary. Thanks!
